pandas/tests/frame/test_block_internals.py

```diff
@@ -644,3 +646,27 @@ def test_to_dict_of_blocks_item_cache():
     assert df.loc[0, "b"] == "foo"
 
     assert df["b"] is ser
+
+
+def test_fillna_sets_valid_block_values():
+    # GH#35731 fillna call was setting Block.values to a Series
+    df = pd.DataFrame(
+        {
+            "a": pd.Series([np.nan, 2.0, 3.0, 1.0]).astype("category"),
+            "b": pd.Series(["A", "A", "B", "C"]).astype("category"),
+            "c": pd.Series(["D", "E", "E", np.nan]).astype("category"),
+        }
+    )
+    cats = df["a"].cat.categories.tolist()
+    # append new category
+    cats.append(0.0)
+    df["a"] = df["a"].astype(pd.CategoricalDtype(categories=cats, ordered=False))
```

Comment:
Can you use the simpler reproducer from https://github.com/pandas-dev/pandas/issues/35731#issuecomment-674409336 ? (that's sufficient for having the "Series as block values" bug)